src/TimeoutMigrationTool/Asp/AspTimeoutsSource.cs

```diff
@@ -0,0 +1,608 @@
+        {
+            this.timeoutTableName = timeoutTableName;
+            this.endpointNameToBelisted = endpointNameToBelisted;
+            this.partitionKeyScope = partitionKeyScope;
+            this.containerName = containerName;
+            this.tablePrefix = tablePrefix;
+            this.batchSize = batchSize;
+            this.connectionString = connectionString;
+        }
+
+        public async Task<IToolState> TryLoadOngoingMigration()
+        {
+            var tableClient = CreateCloudTableClient();
+
+            var toolStateEntity = await TryLoadToolStateEntity(tableClient);
+
+            return toolStateEntity == null ? null : new AspToolState(GetNextBatch, toolStateEntity.RunParameters, toolStateEntity.EndpointName, toolStateEntity.NumberOfBatches);
+        }
+
+        async Task<BatchInfo> GetNextBatch()
+        {
+            var tableClient = CreateCloudTableClient();
+
+            var toolStateEntity = await TryLoadToolStateEntity(tableClient);
+
+            if (toolStateEntity.CurrentBatchNumber > toolStateEntity.BatchNumberAndSizes.Count)
+            {
+                return null;
+            }
+
+            return new BatchInfo(toolStateEntity.CurrentBatchNumber, toolStateEntity.CurrentBatchState,
+                toolStateEntity.BatchNumberAndSizes[toolStateEntity.CurrentBatchNumber - 1].batchSize);
+        }
+
+        async Task<ToolStateEntity> TryLoadToolStateEntity(CloudTableClient tableClient)
+        {
+            var toolStateTable = await GetAndCreateToolStateTableIfNotExists(tableClient);
+
+            var query = new TableQuery<ToolStateEntity>()
+                .Where(
+                    CombineFilters(
+                        GenerateFilterCondition("PartitionKey", QueryComparisons.Equal, ToolStateEntity.FixedPartitionKey),
+                        TableOperators.And,
+                        GenerateFilterConditionForInt("Status", QueryComparisons.Equal, 1)))
+                .Take(1);
+
+            var toolStateEntities = await toolStateTable.ExecuteQuerySegmentedAsync(query, null);
+            var toolStateEntity = toolStateEntities.Results
+                .SingleOrDefault();
+            return toolStateEntity;
+        }
+
+        async Task<CloudTable> GetAndCreateToolStateTableIfNotExists(CloudTableClient tableClient)
+        {
+            var toolStateTable = tableClient.GetTableReference($"{tablePrefix}{AspConstants.ToolStateTableName}");
+            await toolStateTable.CreateIfNotExistsAsync();
+            return toolStateTable;
+        }
+
+        CloudTableClient CreateCloudTableClient()
+        {
+            var account = CloudStorageAccount.Parse(connectionString);
+            var tableClient = account.CreateCloudTableClient();
+            return tableClient;
+        }
+
+        public async Task<IToolState> Prepare(DateTime maxCutoffTime, string endpointName, IDictionary<string, string> runParameters)
```

Comment:
I don't normally care much about long methods, but this one does have a fair bit going on. Can we at least refactor some of the bulkier parts to separate methods?

E.g. the hide and migrate logic could be separate functions each making it a bit easier to understand each.

Reply:
Agreed. The problem was though the decision whether we want to make this even more efficient by introducing concurrency could drastically change the approach we take in this method. That's why I was a bit on the fence about making it beautiful already

Reply:
Refactored